wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+					return 'BLOB';
+				}
+				return $this->translate_sequence( $ast->get_children() );
+			case 'defaultCollation':
+				// @TODO: Check and save in information schema.
+				return null;
+			case 'duplicateAsQueryExpression':
+				// @TODO: How to handle IGNORE/REPLACE?
+
+				// The "AS" keyword is optional in MySQL, but required in SQLite.
+				return 'AS ' . $this->translate( $ast->get_child_node() );
+			case 'indexHint':
+			case 'indexHintList':
+				return null;
+			default:
+				return $this->translate_sequence( $ast->get_children() );
+		}
+	}
+
+	private function translate_token( WP_MySQL_Token $token ) {
+		switch ( $token->id ) {
+			case WP_MySQL_Lexer::EOF:
+				return null;
+			case WP_MySQL_Lexer::AUTO_INCREMENT_SYMBOL:
+				return 'AUTOINCREMENT';
+			case WP_MySQL_Lexer::BINARY_SYMBOL:
+				/*
+				 * There is no "BINARY expr" equivalent in SQLite. We can look for
+				 * the BINARY keyword in particular cases (with REGEXP, LIKE, etc.)
+				 * and then remove it from the translated output here.
+				 */
+				return null;
+			case WP_MySQL_Lexer::SQL_CALC_FOUND_ROWS_SYMBOL:
+				/*
+				 * The "SQL_CALC_FOUND_ROWS" keyword is implemented in the select
+				 * statement translation and then removed from the output here.
+				 */
+				return null;
+			default:
+				return $token->value;
+		}
+	}
+
+	private function translate_sequence( array $nodes, string $separator = ' ' ): ?string {
+		$parts = array();
+		foreach ( $nodes as $node ) {
+			if ( null === $node ) {
+				continue;
+			}
+
+			$translated = $this->translate( $node );
+			if ( null === $translated ) {
+				continue;
+			}
+			$parts[] = $translated;
+		}
+		if ( 0 === count( $parts ) ) {
+			return null;
+		}
+		return implode( $separator, $parts );
+	}
+
+	private function translate_string_literal( WP_Parser_Node $node ): string {
+		$token = $node->get_child_token();
+
+		/*
+		 * 1. Remove bounding quotes.
+		 */
+		$quote = $token->value[0];
+		$value = substr( $token->value, 1, -1 );
+
+		/*
+		 * 2. Normalize escaping of "%" and "_" characters.
+		 *
+		 * MySQL has unusual handling for "\%" and "\_" in all string literals.
+		 * While other sequences follow the C-style escaping ("\?" is "?", etc.),
+		 * "\%" resolves to "\%" and "\_" resolves to "\_" (unlike in C strings).
+		 *
+		 * This means that "\%" behaves like "\\%", and "\_" behaves like "\\_".
+		 * To preserve this behavior, we need to add a second backslash in cases
+		 * where only one is used. To do so correctly, we need to:
+		 *
+		 *  1. Skip all double backslash patterns (as "\\" resolves to "\").
+		 *  2. Add an extra backslash when "\%" or "\_" follows right after.
+		 *
+		 * This may be related to: https://bugs.mysql.com/bug.php?id=84118
+		 */
+		$value = preg_replace( '/(^|[^\\\\](?:\\\\{2}))*(\\\\[%_])/', '$1\\\\$2', $value );
+
+		/*
+		 * 3. Unescape quotes within the string.
+		 */
+		$value = str_replace( $quote . $quote, $quote, $value );
```

Comment:
Is the tokenizer giving us the literal escaped quotes?

Reply:
@adamziel It does give the string to us in full, including escape sequences at the moment. That said, it's a very good question! Should it do so? The tokenizer needs to understand escaping anyway (to know when a literal ends), so it could as well process the escape sequences 🤔

Reply:
> Sounds like we have good reasons to extract the unescaped data from the tokenizer. Do we also have reasons to extract the escaped data?

@adam Actually, I'm not sure if we need the unescaped data. What I meant is we probably could move this logic to the tokenizer because the tokenizer needs to look at escaping anyway. I think I will create a separate ticket to verify escaping more in depth and add this idea as well.